h2/src/test/org/h2/test/db/TestGeneralCommonTableQueries.java

```diff
@@ -469,98 +472,110 @@ private void testColumnNames() throws Exception {
+                "(select 1) union all (select n+1 from r where n < 3)\n"+
+                ")\n"+
+                "select n from r";
+
+        int maxRetries = 3;
+        int expectedNumberOfRows = expectedRowData.length;
+            
+        testRepeatedQueryWithSetup(maxRetries, expectedRowData, expectedColumnNames, expectedNumberOfRows, SETUP_SQL,
+                WITH_QUERY, maxRetries-1, expectedColumnTypes);
+            
+    }    
+    
+    private void testSimple2By4RowRecursiveQuery() throws Exception {
+        
+        String[] expectedRowData =new String[]{"|0|1|10","|1|2|11","|2|3|12","|3|4|13"};
+        String[] expectedColumnTypes =new String[]{"INTEGER","INTEGER","INTEGER"};
+        String[] expectedColumnNames =new String[]{"K","N","N2"};
+        
+        String SETUP_SQL = "-- do nothing";
+        String WITH_QUERY = "with \n"+
+                "r1(n,k) as ((select 1, 0) union all (select n+1,k+1 from r1 where n <= 3)),"+
+                "r2(n,k) as ((select 10,0) union all (select n+1,k+1 from r2 where n <= 13))"+
+                "select r1.k, r1.n, r2.n AS n2 from r1 inner join r2 ON r1.k= r2.k          ";
+
+        int maxRetries = 3;
+        int expectedNumberOfRows = expectedRowData.length;
+            
+        testRepeatedQueryWithSetup(maxRetries, expectedRowData, expectedColumnNames, expectedNumberOfRows, SETUP_SQL,
+                WITH_QUERY, maxRetries-1, expectedColumnTypes);
+            
+    }    
+
+    /*
+        Caused by: java.lang.StackOverflowError
+        at org.h2.result.LazyResult.reset(LazyResult.java:43)
+        at org.h2.command.dml.Select$LazyResultSelect.reset(Select.java:1464)
+        at org.h2.index.ViewIndex.findRecursive(ViewIndex.java:187)
+        at org.h2.index.ViewIndex.find(ViewIndex.java:284)
+        at org.h2.index.ViewIndex.find(ViewIndex.java:161)
+        at org.h2.index.BaseIndex.find(BaseIndex.java:127)
+        at org.h2.index.IndexCursor.find(IndexCursor.java:169)
+        at org.h2.table.TableFilter.next(TableFilter.java:468)
+        at org.h2.command.dml.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1487)
+        at org.h2.result.LazyResult.hasNext(LazyResult.java:79)
+        at org.h2.result.LazyResult.next(LazyResult.java:59)
+        at org.h2.index.ViewCursor.next(ViewCursor.java:49)
+        at org.h2.index.IndexCursor.next(IndexCursor.java:309)
+        at org.h2.table.TableFilter.next(TableFilter.java:499)
+        at org.h2.command.dml.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1487)
+        at org.h2.result.LazyResult.hasNext(LazyResult.java:79)
+        at org.h2.result.LazyResult.next(LazyResult.java:59)
+        at org.h2.index.ViewCursor.next(ViewCursor.java:49)
+        at org.h2.index.IndexCursor.next(IndexCursor.java:309)
+        at org.h2.table.TableFilter.next(TableFilter.java:499)
+        at org.h2.command.dml.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1487)
+     */
+    
```

Comment:
Below is the test case I added to try to re-create the stack overflow scenario. It did not trigger the bad scenario.